### PR TITLE
feat: add plugin capability

### LINKED ISSUE #3: Line numbers
Would it be possible to add support for line numbers?

Bonus points for **not** restricting you to auto-generate them either. For example, when showing a Git diff, the line numbers will change.

### LINKED ISSUE #206: Custom <pre> class name?
In Astro, we'd love to power our new built-in `<Code>` component with Shiki (to eventually replace our current built-in `<Prism>` component). 

However, we would like to keep shikli as an implementation detail, and not leak the `class="shiki"` attribute onto the `<pre>` element if possible. 

Is this something that you / the team is interested in adding as a configuration value to `codeToHtml()` to customize the `<pre>` class name that gets added? If so, I'm happy to provide the PR.

### LINKED ISSUE #330: Support for diffs & code blur
Hi,

It would be great if we was able to highlight diffs similar to this:
1AFD0748-9545-4650-B6DE-C149F83CE5C6

In addition, being able to blur surrounding code so only a section is visible.

Not sure if plug-ins exist to do this.

### LINKED ISSUE #373: Dark mode support mentioned in docs does not work
It requires class with name of theme to use the method below, but the html generated by shiki does not include it
```scss
@media (prefers-color-scheme: light) {
  .shiki.dark-plus {
    display: none;
  }
}
@media (prefers-color-scheme: dark) {
  .shiki.light-plus {
    display: none;
  }
}
```

### LINKED ISSUE #380: Add plugin capability
Shiki would benefit from providing a pluggable plugin architecture.

*Why?*
- Modifications to the current HTML structure are only possible if someone uses `renderHTML` directly and provides their element definition.
- Modifications can be shared, but there are not built-in merge mechanisms if someone wants to apply their own modifications.
- By having plugins it frees project owners/contributors from work/requests - others can built themselves. ;-)
- It builds an ecosystem.

*How?*
Good question. I started to design an (opinionated) approach, you can look at it in #381

### PR DESCRIPTION
**Pull request by muenzpraeger**
- [x] Add a test if possible
- [x] Format all commit messages with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)

- [x] This PR fixes #380, fixes #373, fixes #330, fixes #206, fixes #3
It also supersedes #376.  

See the file `plugin.md` for usage details. I also provided diff plugin as example [here](https://github.com/muenzpraeger/shiki-plugin-diff). Note that the plugin is not published to `npm` yet, so for testing it needs to be linked locally.

### CODE REVIEW

**Comment by orta** [comment]
I'm calling plugin support as above my shiki pay-grade - I'm lightly in favour of this proposal though, I think it's a good way to let us as maintainers not be in the way of folks wanting to do their own interesting things.

Plus, going plugin is literally in my TODO list for every big OSS project I end up maintaining: https://artsy.github.io/blog/2016/07/03/handling-big-projects/

**Comment by muenzpraeger** [comment]
> I think it's a good way to let us as maintainers not be in the way of folks wanting to do their own interesting things.

Exactly [my thinking](https://github.com/shikijs/shiki/issues/380). Been there too often to _not start_ with extensibility in mind - and then you look for the knobs and handles to add/adjust for new functionality that's not part of the core.

The existing functionality (swapping elements or line options) already help a lot. But then they move consumers away from the standard, and eventually future benefits.

Btw - is it @octref's paygrade? ;-)

**Comment by bogdansoare** [comment]
Would love to get this merged and benefit from the ecosystem of plugins.

**Comment by muenzpraeger** [comment]
@octref @orta Thoughts about this PR?

**Comment by octref** [comment]
Hey @muenzpraeger, thanks for your work and sorry for my late response. Has been busy with other things.

Overall I think the PR is good. I'm planning to push for a 1.0 release and break some API, so this could be a good addition. I'll open a new issue for 1.0 plans soon.

Re the PR itself, I want to strike a good balance between easy-to-use and extensibility. Here's a list of goals we should strive for:
- There should be one obvious way to do everything
- The options API should suffice for as much use cases as possible (90%+)
- Core functionalities should be part of the main package (be it done as options or bundled plugins). My thinking on this has changed and I feel diff and dark mode support should be built in.
- The plugin API, with its dynamic hooks, should encapsulate more complex use cases. For example I'd love the effect of https://fatihkalifa.com/typescript-twoslash to be a userland plugin, but diff should be a core plugin. twoslash could be a core plugin that's not included/bundled by default, but can be easily added on.

> It builds an ecosystem.

If done well, yes. I think we got a solid start. Here's the next steps:

- Fix other smaller bugs and merge PRs
- Feature freeze and release a last 0.x version
- Flesh out a better version of the current options API
- Review this proposal in light of the new API

I do want to get this in, but I want to do things in sequence too. Be patient with me :p

File: docs/plugins.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  })
+  .then(highlighter => {
+    highlighter.usePlugin(coolPlugin) // usePlugin() adds the plugin to the highlighter
+    console.log(highlighter.codeToHtml(`console.log('shiki');`, { lang: 'js' }))
+  })
+```
+
+You can also chain multiple plugins. The order of the plugins is important, as they are applied in the order they are passed to `usePlugin()`.
+
+> Note: After invoking `codeToHtml()` the plugins in the highlighter will be reset, so you will need to call `usePlugin()` again if you want to use the plugins again.
```
[the remark this answers is not in the document]
Comment (muenzpraeger) [reply docs/plugins.md:25]: Good question.

When building this I was thinking about the use cases where someone will have multiple code blocks, and where they want different plugins applied to each block.

Like:
- Block 1 - line numbers
- Block 2 - line numbers and blur out some lines
- Block 3 - line numbers and diff highlighting 

So it's mostly really about avoiding conflicting between blocks. The singleton would stay.

Happy to discuss better solutions. 🤞

**Comment by heychazza** [comment]
Exciting stuff, look forward to diff support

**Comment by muenzpraeger** [comment]
@octref I'm all in!

The current Plugin API design is based on that I didn't want to add a breaking change (b/c I'm not the project owner).

I totally agree on that there should be a well-balanced set of built-in functionality. In that regard I'm a strong proponent of having that also built on a plugin architecture, so that it's dogfeeding into the core.

LMK if you want a helping keyboard with v1.

**Comment by LeoDog896** [comment]
Another use case that I've seen is pre/post-processing of shiki. For example, [rehype-pretty-code](https://rehype-pretty-code.netlify.app/) is a plugin for remark that uses shiki for processing. By making it part of the rehype toolchain, it makes it easy to process tokens and other systems (thanks wooorm!)

**Comment by LeoDog896** [comment]
To clarify, I don't think it's a good idea to make an entirely new toolchain for HTML post/preprocessing. I think that encouraging the use of remark would be a good idea for additional HTML processing, while allowing some sort of token processing internally into shiki.

**Pull request closed without merging**